sdk/src/Core/Plugins/EC2Plugin.cs

```diff
@@ -60,36 +66,98 @@ public string ServiceName
         public bool TryGetRuntimeContext(out IDictionary<string, object> context)
         {
             context = null;
+            IDictionary<string, object> dict = new Dictionary<string, object>();
 
-            var dict = new Dictionary<string, object>();
-            if (EC2InstanceMetadata.InstanceId != null)
+            // get the token
+            string token = GetToken();
+
+            // get the metadata
+            dict = GetMetadata(token);
+
+            if (dict == null || dict.Count == 0)
             {
-                dict.Add("instance_id", EC2InstanceMetadata.InstanceId);
+                _logger.DebugFormat("Could not get instance metadata");
+                return false;
             }
 
-            if (EC2InstanceMetadata.AvailabilityZone != null)
+            context = dict;
+            return true;
+        }
+
+        private string GetToken()
+        {
+            string token = null;
+            try
             {
-                dict.Add("availability_zone", EC2InstanceMetadata.AvailabilityZone);
+                Dictionary<string, string> header = new Dictionary<string, string>(1);
+                header.Add("X-aws-ec2-metadata-token-ttl-seconds", "60");
+                token = DoRequest(metadata_base_url + "api/token", HttpMethod.Put, header).Result;
             }
+            catch (Exception)
+            {
+                _logger.DebugFormat("Failed to get token for IMDSv2");
+            }
+
+            return token;
+        }
 
-            if (EC2InstanceMetadata.InstanceType != null)
+
+        private IDictionary<string, object> GetMetadata(string token)
+        {
+            try
             {
-                dict.Add("instance_size", EC2InstanceMetadata.InstanceType);
+                Dictionary<string, string> headers = null;
+                if (token != null)
+                {
+                    headers = new Dictionary<string, string>(1);
+                    headers.Add("X-aws-ec2-metadata-token", token);
+                }
+                string identity_doc_url = metadata_base_url + "dynamic/instance-identity/document";
+                string doc_string = DoRequest(identity_doc_url, HttpMethod.Get, headers).Result;
+                return ParseMetadata(doc_string);
             }
+            catch (Exception)
+            {
+                _logger.DebugFormat("Error occurred while getting EC2 metadata");
+                return null;
```

Comment:
Not much experience with c# but is it possible to return an empty dictionary here instead of null? No points to do a null check at the caller if it's same semantics as empty

https://stackoverflow.com/a/54902869

Reply:
Can't use `ImmutableDictionary` since it's [only supported](https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutabledictionary-2.empty?view=netcore-3.1#moniker-applies-to) for .NetCore and .Net 5. Bummer :(

returning a new empty Dictionary<>() instead to avoid the null check.